app/shared/design-system/components/primitives/numericGestureControl.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * @description
+ * - detects pointer movements in the specified direction
+ * - avails an appropriate cursor
+ * - dispatches {onValueChange} method while pointermove and pointerdown are active
+ * - dispatched {onValueChange} recieves a new value: old value + accumulation of the pointer move axis
+ * @example
+ * numericGestureControl(document.querySelector('input'), {
+ *   onValueChange: (event) => {
+ *     event.preventDefault();
+ *     event.target.value = event.value;
+ *     event.target.select();
+ *   }
+ * });
+ */
+export const numericGestureControl = (
+  targetNode,
+  { initialValue = 0, direction = "horizontal", onValueChange = ({}) => null }
+) => {
+  const { ownerDocument } = targetNode;
+  const eventNames = ["pointerup", "pointerdown", "pointermove"];
+  const state = {
+    value: initialValue,
+    cursor: cursorNode.cloneNode(true),
+    offset: 0,
+    velocity: direction === "horizontal" ? 1 : -1,
+    position: direction === "horizontal" ? "left" : "top",
+    rotation: direction === "horizontal" ? "rotate(0deg)" : "rotate(90deg)",
+    direction: direction === "horizontal" ? "ew-resize" : "ns-resize",
+    dropShadow: `drop-shadow(${
+      direction === "horizontal" ? "0 1px" : "1px 0"
+    } 1.1px rgba(0,0,0,.4))`,
+  };
+  const handleEvent = (event) => {
+    const { type, pressure, offsetX, offsetY, movementY, movementX } = event;
+    const offset = direction === "horizontal" ? offsetX : offsetY;
+    const movement = direction === "horizontal" ? movementX : -movementY;
+    switch (type) {
+      case "pointerup": {
+        ownerDocument.exitPointerLock();
+        break;
+      }
+      case "pointerdown": {
+        const handlePointerLockEvent = () => {
+          const { pointerLockElement } = ownerDocument;
+          switch (ownerDocument.pointerLockElement) {
+            case targetNode: {
+              state.offset = offset;
+              state.cursor.style.position = "absolute";
+              state.cursor.style.top = `${offsetY}px`;
+              state.cursor.style.left = `${offsetX}px`;
+              state.cursor.style.transform = state.rotation;
+              state.cursor.style.setProperty("--drop-shadow", state.dropShadow);
```

Comment:
I feel uneasy looking at those state mutations here. If you wanted to expose styles, then it's okay to put them into the state but I would separate styles and state somehow. See my above comments.

But from what I'm reading here it doesn't even need to be exposed yet (which I'm not sure yet, since I can totally imagine how we'd like to customize styles later when we'll be implementing flex gap etc controls).

Reply:
yeh, its only possible to write like that if you never expose this state to the outside, maybe worth adding a comment?